impl/observable.js

```diff
@@ -0,0 +1,650 @@
+					},
+					error: reject,
+					complete() {
+						resolve(state);
+					},
+				},
+				{
+					signal,
+				},
+			);
+		});
+	}
+	toArray(options) {
+		return this.reduce(
+			(arr, value) => {
+				arr.push(value);
+				return arr;
+			},
+			[],
+			options,
+		);
+	}
+	catch(handleError) {
+		return new Observable((destination) => {
+			this.subscribe(
+				{
+					next(value) {
+						destination.next(value);
+					},
+					error(error) {
+						Observable.from(handleError(error)).subscribe(destination, {
+							signal: destination.signal,
+						});
+					},
+					complete() {
+						destination.complete();
+					},
+				},
+				{
+					signal: destination.signal,
+				},
+			);
+		});
+	}
+	finally(onFinalize) {
+		return new Observable((destination) => {
+			destination.addTeardown(onFinalize);
+			this.subscribe(destination, {
+				signal: destination.signal,
+			});
+		});
+	}
+	[Symbol.asyncIterator]() {
+		let ac;
+		let deferred = [];
+		let buffer = [];
+		let hasError = false;
+		let error = undefined;
+		let isComplete = false;
+		return {
+			next: () => {
+				return new Promise((resolve, reject) => {
+					if (buffer.length > 0) {
+						resolve({ value: buffer.shift(), done: false });
+						return;
+					}
+					if (hasError) {
+						reject(error);
+						return;
+					}
+					if (isComplete) {
+						resolve({ value: undefined, done: true });
+						return;
+					}
+					if (!ac) {
+						ac = new AbortController();
+						this.subscribe(
+							{
+								next(value) {
+									if (deferred.length > 0) {
+										const [resolve] = deferred.shift();
+										resolve({ value, done: false });
+									} else {
+										buffer.push(value);
+									}
+								},
+								error(err) {
+									if (buffer.length > 0) {
+										hasError = true;
+										error = err;
+									} else {
+										while (deferred.length > 0) {
+											const [, reject] = deferred.shift();
+											reject(err);
+										}
+									}
+								},
+								complete() {},
+							},
+							{
+								signal: ac.signal,
+							},
+						);
+					}
+					deferred.push([resolve, reject]);
+				});
+			},
+			throw: (err) => {
+				return new Promise((_resolve, reject) => {
+					ac?.abort();
+					hasError = true;
+					error = err;
+					for (const [, deferredReject] of deferred) {
+						deferredReject(error);
+					}
+					reject(error);
+				});
+			},
+			return: () => {
+				return new Promise((resolve) => {
+					ac?.abort();
+					isComplete = true;
+					for (const [deferredResolve] of deferred) {
+						deferredResolve({ value: undefined, done: true });
+					}
+					resolve({ value: undefined, done: true });
+				});
+			},
+			[Symbol.asyncIterator]() {
+				return this;
+			},
+		};
+	}
+}
+class Subscriber {
+	#active = true;
+	#abortController = new AbortController();
+	#signal;
+	#teardowns;
+	get signal() {
+		return this.#signal;
+	}
+	get isActive() {
+		return this.#active && !this.signal.aborted;
```

Comment:
This doesn't seem to align to the current spec. AFAICT it's possible to see `isActive === false` while also `this.signal.aborted === false`.